config/src/config.rs

```diff
@@ -102,80 +106,94 @@ impl Default for BaseConfig {
     fn default() -> BaseConfig {
         BaseConfig {
             data_dir_path: PathBuf::from("."),
+            role: RoleType::Validator,
             temp_data_dir: None,
         }
     }
 }
 
 impl BaseConfig {
-    /// Constructs a new BaseConfig with an empty temp directory
-    pub fn new(data_dir_path: PathBuf) -> Self {
+    pub fn new(data_dir_path: PathBuf, role: RoleType) -> Self {
         BaseConfig {
             data_dir_path,
+            role,
             temp_data_dir: None,
         }
     }
 }
 
-#[cfg(any(test, feature = "fuzzing"))]
 impl Clone for BaseConfig {
     fn clone(&self) -> Self {
         Self {
             data_dir_path: self.data_dir_path.clone(),
+            role: self.role,
             temp_data_dir: None,
         }
     }
 }
 
+#[derive(Clone, Copy, Debug, Deserialize, Eq, PartialEq, Serialize)]
+#[serde(rename_all = "snake_case")]
+pub enum RoleType {
+    Validator,
+    FullNode,
+}
+
+impl RoleType {
+    pub fn is_validator(&self) -> bool {
+        *self == RoleType::Validator
+    }
+}
+
+impl std::str::FromStr for RoleType {
+    type Err = failure::Error;
```

Comment:
This should not have been a `failure::Error` like I pointed out in my original comment

Reply:
The only other example in our code leveraged failure::Result:
https://github.com/libra/libra/blob/master/types/src/account_address.rs#L202
And this is actually only used in one place, your [comment](https://github.com/libra/libra/pull/1815#discussion_r350871844) was taken as a suggestion.

I'll put up another commit for this.